### ISSUE TITLE
Serving media from non-local/dummy filesystem

### ISSUE THREAD

**Issue by kksharma1618** [anacrolix/dms#75]
[Not a bug. This is a feature request, and a question]

Currently you have to specify local folder from where dms serves media. Is it possible to provide plugin interface in that part of code, so others can use it to serve media from non-fs locations.

This could be used to serve media from remote network locations, or from non-fs systems entirely. For instance we have an app that has folders, playlists, tags, along with video/audio/thumbnails stored in cloud. We would like to use dms to export that data to local dlna. 
So instead of dlna getting folders, items, thumbnails from local fs, it would allow support for plugin which can provide all that data. Developers can write plugin for their own custom uses.

If that is not something you are interested in, can you please point me to the code in your package which is loading that data from local fs? In case we have to fork/modify this for private use.
= I am not familiar with golang, but I think most of fs related code is in dlna/dms/cds.go I can probably figure that part out.

**Comment by anacrolix** [comment]
This would be a welcomed feature. You would need to distil the current filesystem related stuff as an interface to provide for abstracting out the current stuff, and allow for replacing or adding new locations.

**Comment by kksharma1618** [comment]
@anacrolix 
Thanks. Still not sure which route we want to go with. Currently we are using fuse api to mount our structure as a folder, and using dms on top of it. Don't know how stable that is especially on windows.

**Comment by kksharma1618** [comment]
@anacrolix 
Regarding your previous comment "and allow for replacing or adding new locations"
I was checking your ./dlna/dms/cd-service-desc.go file.
You are specifying actions like CreateObject, DestroyObject, UpdateObject, MoveObject, ImportResource. But I dont see any implementation for them in the code. If we develop an interface, would it have to support those features.

**Comment by anacrolix** [comment]
We can use this issue to track the interface we described?

**Comment by kksharma1618** [comment]
Sure. actually I have a working fork at https://github.com/kksharma1618/dms
https://github.com/anacrolix/dms/compare/master...kksharma1618:master
Which is good enough for our particular use. Instead of adding interface, I just added an argument "-contentProviderServer". I am providing url to our own backend there.
Then in dlna/dms/cds_server.go file, I am pulling content from that api.
That apis provides folders/items data as contentProviderServerItem struct. Which I transform into upnp.Object in Browse action.
That api also provides icon/media url. dms just proxies it.

Basically, "contentProviderServer" provides api endpoint with:
/browse?id={objectid}
-- gives array of children for root -> []contentProviderServerItem{}
-- item for media has thumbnail/media url set.

dms proxies thumbnail/media content from the provided url (I dont want to open up my app server on 0.0.0.0)
(dms.go serveCdpProxy)

I am not that familiar with golang, so code is probably pretty bad. But atleast works for our custom case.

**Comment by anacrolix** [comment]
It looks best if you keep that code for your custom case.
